gslib/tests/test_mv.py

```diff
@@ -119,6 +119,80 @@ def test_move_dir_to_bucket(self):
     self.RunGsUtil(['mv', dir_to_move, suri(bucket_uri)])
     self.AssertNObjectsInBucket(bucket_uri, 2)
 
+  def test_move_bucket_objects_with_duplicate_names_intra_bucket(self):
+    """Tests moving multiple top-level items between buckets."""
+    bucket1_uri = self.CreateBucket()
+    self.CreateObject(bucket_uri=bucket1_uri,
+                      object_name='dir1/file.txt',
+                      contents=b'data')
+    self.CreateObject(bucket_uri=bucket1_uri,
+                      object_name='dir2/file.txt',
+                      contents=b'data')
+    self.AssertNObjectsInBucket(bucket1_uri, 2)
+    bucket2_uri = self.CreateBucket()
+    self.AssertNObjectsInBucket(bucket2_uri, 0)
+
+    self.RunGsUtil(['mv', suri(bucket1_uri, '*'), suri(bucket2_uri)])
+
+    self.AssertNObjectsInBucket(bucket1_uri, 0)
+    self.AssertNObjectsInBucket(bucket2_uri, 2)
+
+  def test_move_bucket_objects_with_duplicate_names_to_bucket_subdir(self):
+    """Tests moving multiple top-level items between buckets."""
+    bucket1_uri = self.CreateBucket()
+    self.CreateObject(bucket_uri=bucket1_uri,
+                      object_name='dir1/file.txt',
+                      contents=b'data')
+    self.CreateObject(bucket_uri=bucket1_uri,
+                      object_name='dir2/file.txt',
+                      contents=b'data')
+    self.AssertNObjectsInBucket(bucket1_uri, 2)
+    bucket2_uri = self.CreateBucket()
+    self.AssertNObjectsInBucket(bucket2_uri, 0)
+
+    self.RunGsUtil(['mv', suri(bucket1_uri, '*'), suri(bucket2_uri, 'dir')])
+
+    self.AssertNObjectsInBucket(bucket1_uri, 0)
+    self.AssertNObjectsInBucket(bucket2_uri, 2)
```

Comment:
For this test, I think it's important to check the destination objects path. e.g how can we tell if the destination object is gs://bucket/dir/dir1/file.txt or gs://bucket/dir1/file1.txt

Reply:
This is the focus of `test_naming.py`, specifically `testRecursiveCopyObjsAndFilesToNonExistentBucketSubDir`

Reply:
Agreed, but there is nothing in this test module that tells us that naming gets called. Plus, it's a less toll on whoever is not familiar with the behavior. Tests are great for documenting the behaviors and having the expected output checked will make it easier to understand how the mv behavior is meant to work for complicated naming behavior like these. Also, we check for the output for the cp tests anyway, why not do the same here?

Reply:
>> but there is nothing in this test module that tells us that naming gets called

Not sure what you mean here. E2E tests are normally higher level than making sure underlying functions get called.

>> having the expected output checked will make it easier to understand how the mv behavior is meant to work

Sure, but I'd rather describe what's happening with a code comment than adding `ls` API calls and breaking the pattern of the rest of the file.

>> we check for the output for the cp tests anyway, why not do the same here

The `mv` tests are largely duplicates of functionality  tested in `test_cp.py`. The original authors probably meant for this to be lighter weight because (1) E2E tests are costly to run and (2) `test_cp.py` already does more detailed output analysis. 

I technically didn't need to add anything to `test_mv.py` because we already get  coverage from `test_cp.py`, but I wanted to be thorough. It's pushing it too far to break the lightweight testing pattern of the rest of the file, and I'd rather delete all the tests I added here.

Reply:
> Not sure what you mean here. E2E tests are normally higher level than making sure underlying functions get called.

You mentioned that test_naming is taking care of testing the behavior. I am saying that `test_naming` is testing the naming behavior, but how do we ensure that mv is actually using those naming behavior. We do not have any unittests that check if naming module is getting called.

> Sure, but I'd rather describe what's happening with a code comment than adding ls API calls and breaking the pattern of the rest of the file.

Regarding the API call, `self.AssertNObjectsInBucket(bucket2_uri, 2)` itself makes the `ls` call, so I don't think you are increasing the number of API calls. Also, I don't think we are breaking any pattern, instead we would be enhancing the quality of the test. The key thing that this test is looking for is the behavior of trying to move multiple directories with same file underneath. For this test, I think the naming of the destination file is important and hence it should be tested. That's the whole point of this PR, to avoid naming conflict. In the current state, all we are testing is that the file is getting moved over. There is nothing here that tells me that the naming was handled correctly for mv command specifically.

> The mv tests are largely duplicates of functionality tested in test_cp.py. The original authors probably meant for this to be lighter weight because (1) E2E tests are costly to run and 

Like I said above, checking for count involves an ls call anyway. So I don't think it was light weight in the first place.

> (2) test_cp.py already does more detailed output analysis.

Right. And if we want to rely on that behavior, we should really ensure that mv calls the cp behavior i.e have some unittests that check if copy_helper etc gets called. Basically if A depends on B, and we have tested B, we can be sure that A is tested by ensuring that A is calling B correctly. But in this case, we do not test if mv is calling cp. So there are two ways to test this, either via E2E test or via unittest test and ensuring that correct calls were made to cp. We have chosen to test this via E2E test, and if we are doing that, we should ensure that the end result is correct and not rely on inner knowledge that mv relies on cp.

> I technically didn't need to add anything to test_mv.py because we already get coverage from test_cp.py, but I wanted to be thorough. It's pushing it too far to break the lightweight testing pattern of the rest of the file, and I'd rather delete all the tests I added here.

I appreciate you writing these tests, but like I said, there is no lightweight testing pattern that we are breaking. The number of API calls remain the same with what I am asking here. 

Regarding deleting all these tests, we need tests for mv because that's what led to having this PR in the first place. We need to ensure the behavior of mv in such cases. Remember that mv had a separate if condition originally which we removed. But, like I said, you need not have e2e tests if you can ensure using unittests that mv will call the cp behavior. That is what would actually make the tests lightweight.

All this is my opinion of course, and feel free to bring in a second reviewer. I am happy to approve this if others feel what I am proposing is not really needed.

Reply:
Lightweight as in amount of code and smaller scope. Sorry, I meant to remove the part about using extra resources b/c extra API calls are not being made.

But I think it's definitely breaking a pattern if we start adding `ls` output analysis in a file where all similar tests do not have it (with one exception `test_stdin_args`). I'm not a fan of rewriting redundant code to what I wrote for `test_cp.py`, especially if it breaks a pattern.

Reply:
> I think we should replace this test with a unit test that simply checks that the appropriate methods/functions are called

That sounds like `test_naming` unless you mean mocking functions and calling `assert_called_with(...)` (I don't suggest that b/c it's usually better practice to test a behavior rather than an underlying mechanism). I can try converting these to `test_naming` unit tests if you want.

> There is not really a good way for me to look at all these tests in mv and say with confidence that this behavior has been already tested. In such cases, the test here should ensure that the behavior is tested.

The behavior being tested is that `mv`'ing 2 files results in 2 files, not 1, which would indicate improper merging on name duplicate. I can add more clarification in code if you want.

> there is not really a reason to follow a pattern IMO

I think the pattern's purpose is to keep tests simple and short, which can be done b/c the more complex behavior is tested in `test_cp.py`.

> I don't think adding 3-4 lines of code is going to make the test heavy.

We have to combine the "count objects in bucket" logic from `AssertNObjectsInBucket` with string matching on the output paths. It will be more than 4 lines but not a lot if we implement a `assert_paths` parameter for `AssertNObjectsInBucket`. But this is unnecessary if you want the `test_mv` tests converted to `test_naming` unit tests (there are other `mv` tests in `test_naming`).

Reply:
I was suggesting `assert_called_ with`. The reason being, you mentioned that `test_cp` is already testing for the paths. So if that is the case and if we truly want to avoid the redundancy to make it lightweight, we can simply have a test that checks if mv is calling cp behavior. My point being, there is nothing here or in other tests for mv that tells me that mv calls copy_helper or other naming helpers as expected.  This is the knowledge that you and I have, but it does not reflect in the test code. Of course we can add a comment explaining this, but the more self documenting code we have, the better.

> it's usually better practice to test a behavior rather than an underlying mechanism

Testing the behavior is better, but you have to ensure that you are testing it correctly. If you say that you will partially test this because it is already tested by a dependent module, you kinda have to test that you are calling that module correctly.

But anyway, I don't want to block this PR for this. I will leave it for the second reviewer or other experts to break the tie as I feel we both have been repeating the same points and it's not leading anywhere. Will approve the PR once the other open comment is resolved.

Reply:
As per my last comment, I'm fine trying to convert the `test_mv.py` E2E tests to `test_naming.py` unit tests but just wanted to check if you have objections first

Reply:
I was not suggesting converting this to `test_naming`, but instead adding a unit test in test_mv itself that verifies that naming or copy_helper is called as expected. Adding this to `test_naming` is fine, but I think we would again have the same issue, i.e nothing in mv tests that we are correctly using the "naming" behavior.

Reply:
Also, this is a shortcoming in test_mv itself. I feel in general there should be a test that ensures it calls copy_helper or the cp functionality correctly. So it's not specific to the test you have here.

Reply:
Done, left one test as E2E b/c the unit test framework was having trouble with simulating downloads following the `test_naming.py` pattern. The one left as E2E does check output paths.

Reply:
LGTM

Reply:
Also, thanks for adding the tests!